java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -69,3 +69,59 @@ private class StringTaintPreservingCallable extends TaintPreservingCallable {
     this.hasName(["format", "formatted", "join"]) and arg = [0 .. getNumberOfParameters()]
   }
 }
+
+private class NumberTaintPreservingCallable extends TaintPreservingCallable {
+  int argument;
+
+  NumberTaintPreservingCallable() {
+    this.getDeclaringType().getASupertype*().hasQualifiedName("java.lang", "Number") and
+    (
+      this instanceof Constructor and
+      argument = 0
+      or
+      this.getName().matches(["to%String", "toByteArray", "%Value"]) and
+      argument = -1
+      or
+      this.getName().matches(["parse%", "valueOf%", "to%String", "decode"]) and
+      argument = 0
+    )
+  }
+
+  override predicate returnsTaintFrom(int arg) { arg = argument }
+}
+
+private class StringBuilderTaintPreservingCallable extends TaintPreservingCallable {
+  StringBuilderTaintPreservingCallable() {
+    exists(Class c | c = this.getDeclaringType().getASourceSupertype*() |
+      (
+        c.hasQualifiedName("java.lang", "StringBuilder") or
+        c.hasQualifiedName("java.lang", "StringBuffer") or
+        c.hasQualifiedName("java.io", "StringWriter")
+      ) and
+      (
+        this.hasName(["append", "insert", "replace", "toString"])
+        or
+        this.(Constructor).getParameterType(0) instanceof TypeString and
+        c = this.getDeclaringType()
+      )
+    )
+  }
+
+  override predicate returnsTaintFrom(int arg) {
+    arg = -1
```

Comment:
This should not apply to the constructor case.
```suggestion
    arg = -1 and not this instanceof Constructor
```

Reply:
Note that constructor calls can have qualifiers, e.g. when passing the value of an enclosing instance to a constructor of a nested class.